src/main/scala/MessageWithFireTime.scala

```diff
@@ -4,6 +4,19 @@ import java.util.Date
 /**
   * wrap msg with scheduledFireTime
   */
-final case class MessageWithFireTime(msg: AnyRef, scheduledFireTime:Date)
```

Comment:
Instead of breaking this out into several traits / classes, what are your thoughts on just changing the definition to this:
`final case class MessageWithFireTimes(msg: AnyRef, scheduledFireTime: Date, previousFiringTime: Option[Date] = None, nextFiringTime: Option[Date] = None)` and then always setting all three of those fields?

This should give us all of the same benefits of your initial commits while preserving backwards compatibility too.

EDIT:
Just to clarify, then we can get rid of the extra traits / classes added below.

Reply:
absolutely - the thinking here was being conservative to maintain API compatibility to the preserve the original class `MessageWithFireTime`.  I'm happy to flatten it, but i felt it was more intrusive as my first contribution :)